tutorials/sphinx-tutorials/multiagent_competitive_ddpg.py

```diff
@@ -0,0 +1,992 @@
+# we put them in a sequence. They will not interfere with each other as each group writes and reads keys in different places.
+#
+
+# Put exploration policies from each group in a sequence
+agents_exploration_policy = TensorDictSequential(*exploration_policies.values())
+
+collector = SyncDataCollector(
+    env,
+    agents_exploration_policy,
+    device=device,
+    frames_per_batch=frames_per_batch,
+    total_frames=total_frames,
+)
+
+######################################################################
+# Replay buffer
+# -------------
+#
+# Replay buffers are a common building piece of off-policy RL sota-implementations.
+# There are many types of buffers, in this tutorial we use a basic buffer to store and sample tensordict
+# data randomly.
+#
+
+replay_buffers = {}
+for group, _agents in env.group_map.items():
+    replay_buffer = ReplayBuffer(
+        storage=LazyTensorStorage(
+            memory_size, device=device
+        ),  # We will store up to memory_size multi-agent transitions
+        sampler=RandomSampler(),
+        batch_size=train_batch_size,  # We will sample batches of this size
+    )
+    replay_buffers[group] = replay_buffer
+
+######################################################################
+# Loss function
+# -------------
+#
+# The DDPG loss can be directly imported from TorchRL for convenience using the
+# :class:`~.objectives.DDPGLoss` class. This is the easiest way of utilising DDPG:
+# it hides away the mathematical operations of DDPG and the control flow that
+# goes with it.
+#
+# It is also possible to have different policies for each group.
+#
+losses = {}
+for group, _agents in env.group_map.items():
+    loss_module = DDPGLoss(
+        actor_network=policies[group],  # Use the non-explorative policies
+        value_network=critics[group],
+        delay_value=True,  # Whether to use a target network for the value
+        loss_function="l2",
+    )
+    loss_module.set_keys(
+        state_action_value=(group, "state_action_value"),
+        reward=(group, "reward"),
+        done=(group, "done"),
+        terminated=(group, "terminated"),
+    )
+    loss_module.make_value_estimator(ValueEstimators.TD0, gamma=gamma)
+
+    losses[group] = loss_module
+
+target_updaters = {
+    group: SoftUpdate(loss, tau=polyak_tau) for group, loss in losses.items()
+}
+
+optimizers = {
+    group: {
+        "loss_actor": torch.optim.Adam(
+            loss.actor_network_params.flatten_keys().values(), lr=lr
+        ),
+        "loss_value": torch.optim.Adam(
+            loss.value_network_params.flatten_keys().values(), lr=lr
+        ),
+    }
+    for group, loss in losses.items()
+}
+
+######################################################################
+# Training utils
+# --------------
+#
+# We do have to define two helper functions that we will use in the training loop.
+# They are very simple and do not contain any important logic.
+#
+#
+
+
+def get_excluded_keys(group: str):
```

Comment:
this looks like a piece of code that will be executed many times but could be cached once and for all

Reply:
it is extremely cheap as it just loops among 2 entries of a python dictionary.

Reply:
Sure! Just looks awkward to define that as a separate function in a turorial IMO
When I see a function in a tuto like this I assume it's some complex stuff that must be done on-the-fly

Reply:
i'll simplify this